src/pykeen/nn/weighting.py

```diff
@@ -3,34 +3,89 @@
 """Various edge weighting implementations for R-GCN."""
 
 from abc import abstractmethod
+from typing import Union, Optional
 
 import torch
 from class_resolver import Resolver
 from torch import nn
 
+try:
+    from torch_scatter import scatter_add, scatter_max, scatter_softmax
+except ImportError:
+    print("torch-scatter is not installed, attention aggregation won't work. "
+          "Install it here: https://github.com/rusty1s/pytorch_scatter")
+
 __all__ = [
     "EdgeWeighting",
     "InverseInDegreeEdgeWeighting",
     "InverseOutDegreeEdgeWeighting",
     "SymmetricEdgeWeighting",
+    "AttentionEdgeWeighting",
     "edge_weight_resolver",
 ]
 
+def softmax(
+    src: torch.Tensor,
+    index: torch.LongTensor,
+    num_nodes: Union[None, int, torch.Tensor] = None,
+    dim: int = 0,
+) -> torch.Tensor:
+    r"""Computes a sparsely evaluated softmax.
+    Given a value tensor :attr:`src`, this function first groups the values
+    along the given dimension based on the indices specified in :attr:`index`,
+    and then proceeds to compute the softmax individually for each group.
+
+    :param src:
+        The source tensor.
+    :param index:
+        The indices of elements for applying the softmax.
+    :param num_nodes:
+        The number of nodes, i.e., :obj:`max_val + 1` of :attr:`index`. (default: :obj:`None`)
+    :param dim:
+        The dimension along which to compute the softmax.
+
+    :return:
+        The softmax-ed tensor.
+    """
+
+    num_nodes = num_nodes or index.max() + 1
+    out = src.transpose(dim, 0)
+    out = out - scatter_max(out, index, dim=0, dim_size=num_nodes)[0][index]
+    out = out.exp()
+    out = out / scatter_add(out, index, dim=0, dim_size=num_nodes)[index].clamp_min(1.0e-16)
+    return out.transpose(0, dim)
 
 class EdgeWeighting(nn.Module):
     """Base class for edge weightings."""
 
+    def __init__(
+            self,
+            **kwargs
+    ):
+        super().__init__()
+        pass
+
     @abstractmethod
-    def forward(self, source: torch.LongTensor, target: torch.LongTensor) -> torch.FloatTensor:
+    def forward(
```

Comment:
goes beyond this PR, but we could consider switching to https://pytorch-geometric.readthedocs.io/en/latest/notes/create_gnn.html#the-messagepassing-base-class  - PyG and its dependencies have become much easier to install as it was when we build these blocks.

Reply:
related https://github.com/pykeen/pykeen/pull/620

Reply:
Generally, I'd support outsourcing the message passing functionality to PyG!